src/vsr/client.zig

```diff
@@ -19,21 +19,130 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
         pub const StateMachine = StateMachine_;
 
         pub const Error = error{
+            BatchSizeZero,
+            TooMuchData,
             TooManyOutstandingRequests,
         };
 
-        pub const Request = struct {
-            pub const Callback = fn (
-                user_data: u128,
-                operation: StateMachine.Operation,
-                results: Error![]const u8,
-            ) void;
+        pub const Callback = fn (
             user_data: u128,
-            // Null iff operation=register.
-            callback: ?Callback,
+            operation: StateMachine.Operation,
+            results: Error![]const u8,
+        ) void;
+
+        pub const Request = struct {
+            // Null iif operation=register.
+            demux_list: ?DemuxList,
```

Comment:
Rather than a pool of Demux, and a linked list on each request, why not just put a BoundedArray on each Request? I think that would save a little memory since we wouldn't need the `next` pointers anymore.

Reply:
Do you mean by not having an adjustable upper bound? If we choose to allow up to 8191 logical batches for each physical request, it will save us ~2MB of memory.
On the other hand, a pool + linked list can be helpful if we choose to share a limited number of logical batches across all requests (let's say, something like 10k logical batches would require only 400KB of memory).

Reply:
:+1: Good point, an adjustable/configurable upper-bound is better, since clients can tune it based on their own average logical batch size.